### PR TITLE
v0 program enrollment

### PR DESCRIPTION
**Pull request by zacharis278**

### CODE REVIEW

File: registrar/apps/api/serializers.py [hunk 1]
```diff
@@ -43,6 +43,17 @@ class LearnerProgramEnrollmentSerializer(serializers.Serializer):
     status = serializers.CharField()
 
 
+class RequestedLearnerProgramEnrollmentSerializer(serializers.Serializer):
```
Comment (schenedx) [inline registrar/apps/api/serializers.py:46]: Can we have this serializer to be in a different file than the serializer used for v1 versions? Or better, just write a new serializer file in the v0 folder? I would like them to be separate

**Review by schenedx — APPROVED** [review]
Besides the minor serializer suggestion, rest LGTM

File: registrar/apps/api/serializers.py (hunk 1 above)
Comment (zacharis278) [reply registrar/apps/api/serializers.py:46]: Humm so do we want to duplicate every serializer or just on an as-needed basis?  This isn't the only one we use, I'd be worried we're unnecessarily repeating ourselves.

Or maybe just do it for input serializers like this and leave the models as-is?

Comment (schenedx) [reply registrar/apps/api/serializers.py:46]: Ah, I didn't realize this serializer could be used for V1 as well. In that case, you can ignore my comment above

File: registrar/apps/api/v0/views.py [hunk 2]
```diff
@@ -114,3 +125,66 @@ def get_queryset(self):
...
+     * 413: Payload too large, over 25 students supplied.
+     * 422: Invalid request, unable to enroll students.
+    """
+    authentication_classes = (JwtAuthentication, SessionAuthentication)
+    permission_classes = (IsAuthenticated,)
+    serializer_class = RequestedLearnerProgramEnrollmentSerializer
+
+    def post(self, request, *args, **kwargs):
+        """ Enroll up to 25 students in program """
+        if not self.program.managing_organization.metadata_readable:
```
Comment (kdmccormick) [inline registrar/apps/api/v0/views.py:151]: Should be checking against `.enrollments_writeable` here.

File: registrar/apps/api/v0/views.py [hunk 3]
```diff
@@ -114,3 +125,66 @@ def get_queryset(self):
...
+
+        response = {}
+        error_status = None
+
+        if not isinstance(request.data, list):
+            raise ValidationError()
+
+        if len(request.data) > 25:
+            return Response(
+                'enrollement limit 25', HTTP_413_REQUEST_ENTITY_TOO_LARGE
```
Comment (kdmccormick) [inline registrar/apps/api/v0/views.py:162]: `enrollment` (sp.)

File: registrar/apps/api/v0/views.py [hunk 4]
```diff
@@ -114,3 +125,66 @@ def get_queryset(self):
...
+    permission_classes = (IsAuthenticated,)
+    serializer_class = RequestedLearnerProgramEnrollmentSerializer
+
+    def post(self, request, *args, **kwargs):
+        """ Enroll up to 25 students in program """
+        if not self.program.managing_organization.metadata_readable:
+            raise PermissionDenied()
+
+        response = {}
+        error_status = None
```
Comment (kdmccormick) [inline registrar/apps/api/v0/views.py:155]: To get into the weeds of the API spec for a bit: 
* `error_status` should be `422` if no students are successfully recorded. 
* `error_status` should be `207` if some are successfully recorded
* `error_status` should be `200` if some all are successfully recorded

In this implementation, `error_status` will be `207` if all of the students are given `invalid-status`, whereas the spec says it should be `422`.

File: registrar/apps/api/serializers.py (hunk 1 above)
Comment (kdmccormick) [inline registrar/apps/api/serializers.py:46]: I suggest `ProgramEnrollmentRequestSerializer`. Rolls off the tongue a little easier.

File: registrar/apps/api/v0/views.py [hunk 5]
```diff
@@ -114,3 +125,66 @@ def get_queryset(self):
...
+                    response[student_key] = 'duplicated'
+                    error_status = HTTP_207_MULTI_STATUS
+                else:
+                    response[student_key] = enrollee['status']
+            else:
+                try:
+                    if 'status' in enrollee_serializer.errors:
+                        response[enrollee['student_key']] = 'invalid-status'
+                    else:
+                        response[enrollee['student_key']] = 'internal-error'
```
Comment (kdmccormick) [inline registrar/apps/api/v0/views.py:183]: I think missing `email` should raise a 422, just like missing `student_key` would. `'internal-error'` is for unexpected processing errors on our side.

**Pull request merged**